selection/randomized/query.py

```diff
@@ -1838,3 +1863,117 @@ def _weights(self, candidate):
+        nactive = targeted_sampler.observed_target_state.shape[0]
+
+        self._normal_sample = np.random.multivariate_normal(mean=np.zeros(nactive), cov=targeted_sampler.target_cov, size =(sample.shape[0]))
+        print(self._normal_sample.shape)
+        self._sample = np.concatenate((sample, np.tile(self.observed, (sample.shape[0], 1))), axis=1)
+        self._logden = targeted_sampler.log_randomization_density(self._sample)
+        self._delta = np.concatenate((sample, self._normal_sample), axis=1)
+
+
+    def pivot(self,
+              linear_func,
+              candidate,
+              alternative='twosided'):
+        '''
+        alternative : ['greater', 'less', 'twosided']
+            What alternative to use.
+        Returns
+        -------
+        pvalue : np.float
+        '''
+
+        if alternative not in ['greater', 'less', 'twosided']:
+            raise ValueError("alternative should be one of ['greater', 'less', 'twosided']")
+
+        observed_stat = self.targeted_sampler.observed_target_state.dot(linear_func)
+
+        candidate_sample, weights = self._weights(linear_func, candidate)
+        #print("candidate", candidate)
+        sample_stat = np.array([linear_func.dot(s) for s in candidate_sample[:, self.targeted_sampler.target_slice]])
+
+        pivot = np.mean((sample_stat <= observed_stat) * weights) / np.mean(weights)
+
+        if alternative == 'twosided':
+            return 2 * min(pivot, 1 - pivot)
+        elif alternative == 'less':
+            return pivot
+        else:
+            return 1 - pivot
+
+    def confidence_interval(self, linear_func, level=0.90, how_many_sd=20):
+
+        target_delta = self._delta[:,self.targeted_sampler.target_slice]
+        projected_delta = target_delta.dot(linear_func)
+        projected_observed = self.observed.dot(linear_func)
+        std_projected_delta = np.sqrt(np.dot(linear_func.T, self.targeted_sampler.target_cov).dot(linear_func))
+
+        delta_min, delta_max = projected_delta.min(), projected_delta.max()
+
+        _norm = np.linalg.norm(linear_func)
+        grid_min, grid_max = -how_many_sd * np.std(projected_delta), how_many_sd * np.std(projected_delta)
+        print("grid", grid_min, grid_max)
+
+        def _rootU(gamma):
+            return self.pivot(linear_func,
+                              projected_observed + gamma,
+                              alternative='less') - (1 - level) / 2.
+        def _rootL(gamma):
+            return self.pivot(linear_func,
+                              projected_observed + gamma,
+                              alternative='less') - (1 + level) / 2.
+
+        upper = bisect(_rootU, grid_min, grid_max, xtol=1.e-5*(grid_max - grid_min))
+        lower = bisect(_rootL, grid_min, grid_max, xtol=1.e-5*(grid_max - grid_min))
+
+        return lower + projected_observed, upper + projected_observed
+
+    # Private methods
+
+    def _weights(self, linear_func, candidate):
+
```

Comment:
Shouldn't we see some cross terms in the covariance somewhere?

Reply:
We see them through target_sampler.log_randomization_density(), which knows about the decomposition. Then we can use the sample consisting of optimization variables and the target only as arguments of target_sampler.log_randomization_density().

Reply:
I'll have to go through the calculations, but I would have expected residual matrix to involve cross terms of the covariance...

Reply:
This class is just decomposing the target in terms of target.dot(keep) which 1D direction of interest. The score has been decomposed in terms of the whole target vector $\bar{\beta}_E$ earlier in target_sampler.setup_target(). So the cross-terms of the covariance are used before we call this opt_weighted_intervals class.